### PR TITLE
 Declaration emit fixes for binding pattern in variable statements 

### PR DESCRIPTION
**Pull request by sheetalkamat**
Handles #2023

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -1581,8 +1581,15 @@ module ts {
 
             function determineIfDeclarationIsVisible() {
                 switch (node.kind) {
-                    case SyntaxKind.VariableDeclaration:
                     case SyntaxKind.BindingElement:
+                        return isDeclarationVisible(<Declaration>node.parent.parent);
+                    case SyntaxKind.VariableDeclaration:
+                        if (isBindingPattern(node.name) &&
+                            !(<BindingPattern>node.name).elements.length) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: I'd prefer you just compared to `0`.

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -1024,62 +1024,94 @@ module ts {
...
+                function getBindingElementTypeVisibilityError(symbolAccesibilityResult: SymbolAccessiblityResult): SymbolAccessibilityDiagnostic {
+                    var diagnosticMessage = getVariableDeclarationTypeVisibilityDiagnosticMessage(symbolAccesibilityResult);
+                    return diagnosticMessage !== undefined ? {
+                        diagnosticMessage,
+                        errorNode: bindingElement,
+                        typeName: bindingElement.name
+                    } : undefined;
+                }
+
+                if (bindingElement.name) {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: When would the `name` be `undefined`?

Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Never mind, I see that the reason is that you can have an omitted element such as the following:

`var [a, , c] = foo();`

But please put a comment with an example.

Also, I believe that @JsonFreeman mentioned that for a well-formed tree, it is more appropriate to check the syntax kind.

Comment (JsonFreeman) [inline src/compiler/emitter.ts]: Namely, when bindingElement.name is undefined, it's because the syntax kind is OmittedExpression and not BindingElement

**Pull request merged**